File: android/app/src/main/java/com/fyrbyadditive/famesmartblinds/data/model/BlindDevice.kt
package com.fyrbyadditive.famesmartblinds.data.model

import android.bluetooth.BluetoothDevice
import com.fyrbyadditive.famesmartblinds.data.remote.CalibrationStatusResponse
import com.fyrbyadditive.famesmartblinds.data.remote.DeviceStatus
import java.util.Date
import java.util.UUID

/**
 * Represents a FAME Smart Blinds device.
 *
 * Architecture note: BLE is only used during initial device setup (when device has no WiFi).
 * Once WiFi is configured, the device disables BLE advertising and all communication
 * (control, configuration, status) happens over HTTP via mDNS discovery.
 */
data class BlindDevice(
    val id: UUID = UUID.randomUUID(),

    // BLE properties - only used during initial setup before WiFi is configured
    val bluetoothDevice: BluetoothDevice? = null,
    val rssi: Int = 0,

    // Device info
    val name: String = "Unknown Device",
    val deviceId: String = "",  // 8-char hex ID from MAC address, used as stable identifier
    val macAddress: String = "",

    // Configuration (readable via BLE during setup)
    val wifiSsid: String = "",
    val mqttBroker: String = "",

    // Connection state
    val bleConnected: Boolean = false,  // Only true during setup
    val wifiConnected: Boolean = false,
    val mqttConnected: Boolean = false,

    // Network info (when WiFi connected - primary communication method post-setup)
    val ipAddress: String? = null,

    // Blind state
    val state: BlindState = BlindState.UNKNOWN,
    val position: Int = 0,

    // Calibration state
    val isCalibrated: Boolean = false,
    val cumulativePosition: Int = 0,
    val maxPosition: Int = 0,
    val calibrationState: String = "idle",

    // Status string from device
    val statusString: String = "",

    // Timestamps
    val lastSeen: Date = Date()
) {
    /**
     * Whether this device can be controlled via HTTP (has WiFi connection)
     */
    val canControlViaHttp: Boolean
        get() = ipAddress != null

    /**
     * Whether this device needs setup (found via BLE but no WiFi)
     */
    val needsSetup: Boolean
        get() = ipAddress == null && bluetoothDevice != null

    /**
     * Update device from status string like "wifi:192.168.1.100,mqtt:ok,servo:ok"
     */
    fun updateFromStatus(status: String): BlindDevice {
        var device = this.copy(statusString = status)

        status.split(",").forEach { component ->
            val parts = component.split(":")
            if (parts.size >= 2) {
                val key = parts[0]
                val value = parts[1]

                device = when (key) {
                    "wifi" -> {
                        when {
                            value == "connecting" -> device.copy(wifiConnected = false)
                            value == "disconnected" -> device.copy(wifiConnected = false, ipAddress = null)
                            else -> device.copy(wifiConnected = true, ipAddress = value)
                        }
                    }
                    "mqtt" -> device.copy(mqttConnected = value == "ok")
                    else -> device
                }
            }
        }

        return device
    }

    /**
     * Update device from HTTP status response
     */
    fun updateFromDeviceStatus(status: DeviceStatus): BlindDevice {
        return copy(
            state = BlindState.fromString(status.state),
            position = status.position,
            isCalibrated = status.calibration?.calibrated ?: false,
            cumulativePosition = status.calibration?.cumulativePosition ?: 0,
            maxPosition = status.calibration?.maxPosition ?: 0,
            calibrationState = status.calibration?.state ?: "idle"
        )
    }

    /**
     * Update device from calibration status response
     */
    fun updateFromCalibrationStatus(status: CalibrationStatusResponse): BlindDevice {
        return copy(
            isCalibrated = status.calibrated,
            cumulativePosition = status.position,
            maxPosition = status.maxPosition,
            calibrationState = status.calibrationState
        )
    }

    companion object {
        /**
         * Extract device ID from a name like "FAMEBlinds_23c57e80"
         */
        fun extractDeviceId(name: String?): String {
            if (name == null) return ""

            val underscoreIndex = name.lastIndexOf('_')
            if (underscoreIndex != -1) {
                val suffix = name.substring(underscoreIndex + 1)
                // Validate it looks like a hex device ID (8 hex chars)
                if (suffix.length == 8 && suffix.all { it.isDigit() || it in 'a'..'f' || it in 'A'..'F' }) {
                    return suffix.lowercase()
                }
            }
            return ""
        }

        /**
         * Create from BLE discovery - used during initial device setup only
         */
        fun fromBle(
            bluetoothDevice: BluetoothDevice,
            rssi: Int = 0,
            advertisedName: String? = null
        ): BlindDevice {
            val name = advertisedName ?: bluetoothDevice.name ?: "Unknown Device"
            return BlindDevice(
                bluetoothDevice = bluetoothDevice,
                rssi = rssi,
                name = name,
                deviceId = extractDeviceId(name),
                lastSeen = Date()
            )
        }

        /**
         * Create from mDNS/HTTP discovery - primary discovery method for configured devices
         */
        fun fromHttp(
            name: String,
            ipAddress: String,
            deviceId: String = ""
        ): BlindDevice {
            return BlindDevice(
                name = name,
                ipAddress = ipAddress,
                deviceId = deviceId,
                wifiConnected = true,
                lastSeen = Date()
            )
        }
    }
}

File: ios/FAME Smart Blinds/Models/BlindDevice.swift
import Foundation
import CoreBluetooth

/// Represents a FAME Smart Blinds device.
///
/// Architecture note: BLE is only used during initial device setup (when device has no WiFi).
/// Once WiFi is configured, the device disables BLE advertising and all communication
/// (control, configuration, status) happens over HTTP via mDNS discovery.
@MainActor
class BlindDevice: Identifiable, ObservableObject {
    let id: UUID

    // BLE properties - only used during initial setup before WiFi is configured
    var peripheral: CBPeripheral?
    var rssi: Int

    // Device info
    @Published var name: String
    @Published var deviceId: String  // 8-char hex ID from MAC address, used as stable identifier
    @Published var macAddress: String

    // Configuration (readable via BLE during setup)
    @Published var wifiSsid: String
    @Published var mqttBroker: String

    // Connection state
    @Published var bleConnected: Bool = false  // Only true during setup
    @Published var wifiConnected: Bool = false
    @Published var mqttConnected: Bool = false

    // Network info (when WiFi connected - primary communication method post-setup)
    @Published var ipAddress: String?

    // Blind state
    @Published var state: BlindState = .unknown
    @Published var position: Int = 0

    // Calibration state
    @Published var isCalibrated: Bool = false
    @Published var cumulativePosition: Int = 0
    @Published var maxPosition: Int = 0
    @Published var calibrationState: String = "idle"

    // Status string from device
    @Published var statusString: String = ""

    // Timestamps
    var lastSeen: Date

    /// Initialize from BLE discovery - used during initial device setup only
    init(peripheral: CBPeripheral, rssi: Int = 0, advertisedName: String? = nil) {
        self.id = peripheral.identifier
        self.peripheral = peripheral
        self.rssi = rssi
        // Use advertised name if provided (not cached), otherwise fall back to peripheral.name
        let actualName = advertisedName ?? peripheral.name ?? "Unknown Device"
        self.name = actualName
        // Extract device ID from name (e.g., "FAMEBlinds_23c57e80" -> "23c57e80")
        self.deviceId = BlindDevice.extractDeviceId(from: actualName)
        self.macAddress = ""
        self.wifiSsid = ""
        self.mqttBroker = ""
        self.lastSeen = Date()
    }

    /// Extract device ID from a name like "FAMEBlinds_23c57e80"
    static func extractDeviceId(from name: String?) -> String {
        guard let name = name else { return "" }
        // Look for underscore followed by hex ID (8 characters)
        if let underscoreIndex = name.lastIndex(of: "_") {
            let suffix = String(name[name.index(after: underscoreIndex)...])
            // Validate it looks like a hex device ID (8 hex chars)
            if suffix.count == 8 && suffix.allSatisfy({ $0.isHexDigit }) {
                return suffix.lowercased()
            }
        }
        return ""
    }

    /// Initialize from mDNS/HTTP discovery - primary discovery method for configured devices
    init(name: String, ipAddress: String, deviceId: String = "") {
        self.id = UUID()
        self.peripheral = nil
        self.rssi = 0
        self.name = name
        self.deviceId = deviceId
        self.macAddress = ""
        self.wifiSsid = ""
        self.mqttBroker = ""
        self.ipAddress = ipAddress
        self.wifiConnected = true
        self.lastSeen = Date()
    }

    func updateFromStatus(_ status: String) {
        statusString = status

        // Parse status string like "wifi:192.168.1.100,mqtt:ok,servo:ok"
        let components = status.split(separator: ",")
        for component in components {
            let parts = component.split(separator: ":")
            guard parts.count >= 2 else { continue }

            let key = String(parts[0])
            let value = String(parts[1])

            switch key {
            case "wifi":
                if value == "connecting" {
                    wifiConnected = false
                } else if value == "disconnected" {
                    wifiConnected = false
                    ipAddress = nil
                } else {
                    wifiConnected = true
                    ipAddress = value
                }
            case "mqtt":
                mqttConnected = value == "ok"
            default:
                break
            }
        }
    }

    func updateFromDeviceStatus(_ deviceStatus: DeviceStatus) {
        self.state = BlindState(rawValue: deviceStatus.state) ?? .unknown
        self.position = deviceStatus.position

        if let calibration = deviceStatus.calibration {
            self.isCalibrated = calibration.calibrated
            self.cumulativePosition = calibration.cumulativePosition
            self.maxPosition = calibration.maxPosition
            self.calibrationState = calibration.state
        }
    }

    func updateFromCalibrationStatus(_ status: CalibrationStatusResponse) {
        self.isCalibrated = status.calibrated
        self.cumulativePosition = status.position
        self.maxPosition = status.maxPosition
        self.calibrationState = status.calibrationState
    }
}

extension BlindDevice: Equatable {
    nonisolated static func == (lhs: BlindDevice, rhs: BlindDevice) -> Bool {
        lhs.id == rhs.id
    }
}

extension BlindDevice: Hashable {
    nonisolated func hash(into hasher: inout Hasher) {
        hasher.combine(id)
    }
}
